hydrolib/core/io/ini/models.py

```diff
@@ -285,6 +286,53 @@ def validate(cls, v):
                 )
         return super().validate(v)
 
+    @staticmethod
+    def _get_friction_root_validator(
+        frictionid_attr: str,
+        frictiontype_attr: str,
+        frictionvalue_attr: str,
+    ):
+        """
+        Make a root_validator that verifies whether the crosssection definition (subclass)
+        has a valid friction specification.
+        Supposed to be embedded in subclasses for their friction fields.
+
+        Args:
+            frictionid_attr: name of the frictionid attribute in the subclass.
+            frictiontype_attr: name of the frictiontype attribute in the subclass.
+            frictionvalue_attr: name of the frictionvalue attribute in the subclass.
+
+        Returns:
+            root_validator: to be embedded in the subclass that needs it.
+        """
+
+        def validate_friction_specification(cls, values):
+            """
+            The actual validator function.
+
+            Args:
+            cls: The subclass for which the root_validator is called.
+            values (dict): Dictionary of values to create a CrossSectionDefinition subclass.
+            """
+            frictionid = values.get(frictionid_attr) or ""
+            frictiontype = values.get(frictiontype_attr) or ""
+            frictionvalue = values.get(frictionvalue_attr) or ""
+
+            if frictionid == "":
+                if frictiontype == "" or frictionvalue == "":
+                    raise ValueError(
+                        f"Cross section with id \"{values.get('id', '')}\" is missing any friction specification."
+                    )
+            else:
+                if frictiontype != "" or frictionvalue != "":
+                    raise ValueError(
+                        f"Cross section with id \"{values.get('id', '')}\" has duplicate friction specification (both {frictionid_attr} and {frictiontype_attr}/{frictionvalue_attr})."
+                    )
+
+            return values
```

Comment:
It is good practice to override the `BaseModel_get_identifier` method for each identifiable derived BaseModel class. This way, when **any** ValidationError is raised for this class, the identifier will be included in the resulting error message, which makes it easier for the user to identify problems with their model. 
In this case you have added the id in the error message for this specific validation, which is good, but for other validation errors the id will not be included. 
Please check for example `Boundary._get_identifier` in hydrolib.core.io.ext.models.py

Reply:
Cool! I have now added this subroutine to ```CrossSectionDefinition``` in an upcoming commit.
However, I cannot use this inside the root_validator as created by ```def make_list_length_root_validator```, as that one does not have access to the current object (it only receives the static class via ```cls```).